library/std/src/sys/pal/windows/api.rs

```diff
@@ -34,6 +34,97 @@ use core::ptr::addr_of;
 
 use super::c;
 
+/// Creates a null-terminated UTF-16 string from a str.
+macro_rules! wide_str {
+    ($str:expr) => {
+        utf16!(concat!($str, '\0'))
+    };
+}
+
+/// Creates a UTF-16 string from a str without null termination.
+macro_rules! utf16 {
+    // Note: this macro uses triple underscores to avoid const cycles
+    ($str:expr) => {{
+        const ___UTF8: &str = $str;
+        const ___UTF16_LEN: usize = crate::sys::pal::windows::api::utf16_len(___UTF8);
+        const ___UTF16: [u16; ___UTF16_LEN] = crate::sys::pal::windows::api::to_utf16(___UTF8);
+        &___UTF16
+    }};
+}
+
+/// Gets the UTF-16 length of a UTF-8 string, for use in the wide_str macro.
+pub const fn utf16_len(s: &str) -> usize {
+    let s = s.as_bytes();
+    let mut i = 0;
+    let mut len = 0;
+    while i < s.len() {
+        // the length of a UTF-8 encoded code-point is given by the number of
+        // leading ones, except in the case of ASCII.
+        let utf8_len = match s[i].leading_ones() {
+            0 => 1,
+            n => n as usize,
+        };
+        i += utf8_len;
+        len += if utf8_len < 4 { 1 } else { 2 };
+    }
+    len
+}
+
+/// Const convert UTF-8 to UTF-16, for use in the wide_str macro.
+///
+/// Note that this is designed for use in const contexts so is not optimized.
+pub const fn to_utf16<const UTF16_LEN: usize>(s: &str) -> [u16; UTF16_LEN] {
```

Comment:
Do you think if this needs a test? This feels a bit non-trivial.

Reply:
Yeah, would be best. Also I should probably add a null check somewhere.

Reply:
I've added a few test cases that should exercise the different code paths. Though in practice I suspect it won't get much use beyond ASCII. Well, possibly a bit of latin-1.